Implement mongo 3.4 Decimal128 processing in meteor

zebra-lucky opened the pull request:
Fix `colleciton.insert({field: Decimal128})`.

zebra-lucky commented:
Changed PR desrciption, there is problesms to implement Decimal128 support in meteor.
Need to disscussion if someone interesting.

hwillson commented on packages/mongo/mongo_driver_tests.js:
```diff
@@ -0,0 +1,9 @@
+const MongoDB = NpmModuleMongodb;
```
Instead of accessing `MongoDB` this way, could you follow the approach that's currently used with some of the [other tests](https://github.com/meteor/meteor/blob/5e9ff342a8ce9e488e2af4086d39ac1e73325e9c/packages/mongo/mongo_livedata_tests.js#L3237): `const MongoDB = MongoInternals.NpmModules.mongodb.module;` This helps keep things more uniform.

hwillson commented on packages/mongo/mongo_driver_tests.js:
```diff
@@ -0,0 +1,9 @@
+const MongoDB = NpmModuleMongodb;
+
+
+Tinytest.add('mongo_driver - insert Decimal128', function (test) {
```
Thanks for adding this test. Instead of creating a new `mongo_driver_tests.js` file however, could you add this test to the existing `mongo_livedata_tests.js` file? That's where other tests like this one live.

hwillson commented on packages/mongo/package.js:
```diff
@@ -80,6 +80,7 @@ Package.onUse(function (api) {
 });
 
 Package.onTest(function (api) {
+  api.use('npm-mongo', 'server');
```
If you switch to using the `MongoInternals.NpmModules.mongodb.module` approach I mentioned earlier, this shouldn't be needed.

hwillson commented on packages/mongo/package.js:
```diff
@@ -88,6 +89,7 @@ Package.onTest(function (api) {
   // XXX test order dependency: the allow_tests "partial allow" test
   // fails if it is run before mongo_livedata_tests.
   api.addFiles('mongo_livedata_tests.js', ['client', 'server']);
+  api.addFiles('mongo_driver_tests.js', ['server']);
```
As mentioned above, please move this test into `mongo_livedata_tests.js`.

zebra-lucky replied:
Hello!

I was found additional problems If adding `Decimal128` type this way.
Can't add support to `minimongo`. I'm try using npm package `bson`
to add new EJSON type, but different imports leaves problems.

I will try to make it works, and then change PR code with your remarks.

zebra-lucky replied:
There is still one problem with decimal types. There is `decimal.js` or `big.js` libraries
to make calculations with decimal precision. Maybe should make direct maping to one of these,
instead using `bson` type, which can make only toString conversion.

Need some time to investigate if it possible (use `Decimal128` in `mongodb` and say `big.js` in `Meteor` code).

zebra-lucky replied:
Maybe I should add separate package to this PR to support Decimal128 type, something like `mongo-id`.

zebra-lucky replied:
Not found implemetation of IEEE 754-2008 decimal128 in javascript, but there is [decimal.js](https://github.com/MikeMcl/decimal.js/), which can be a workaround.

Need to add tests and will make commit in a day or two.

zebra-lucky replied:
Hello! I'm removed tests from `mongo` package completely, instead will add tests to new `mongo-decimal` package today.

zebra-lucky commented:
Changed code in mongo_driver.js, removed stalled tests.
Added `mongo-decimal` package, yet need to add tests to this package.
Rebased and squashed to one commit.

hwillson commented:
Hi @zebra-lucky - sorry for the large delay in getting back to you here. I think we're just about ready to get this merged, but would you mind making one small adjustment? Could you move the `mongo-decimal` package from its current location, to `/packages/non-core/mongo-decimal`? Thanks!

zebra-lucky commented:
Hi! I'll move package in a half a hour. Should I rebase PR on current master?

zebra-lucky commented:
Moved and rebased to current devel

zebra-lucky commented:
Failed some test ```S: tinytest - observeChanges - unordered - basics : !!!!!!!!! FAIL !!!!!!!!!!!```

hwillson commented:
Excellent, thanks @zebra-lucky! I've restarted the test run so we'll see what happens in a bit. That test is known to sometimes fail randomly.

hwillson commented:
Looks like we're all set - LGTM!

thearabbit commented:
Excuse me, Have any document about this?

zebra-lucky commented:
https://mikemcl.github.io/decimal.js/
https://github.com/MikeMcl/decimal.js/

thearabbit commented:
Could you example for `DB inset/find...`
```js
import {Decimal} from 'decimal.js'
...
let score = new Decimal('123.4567')
Collection.insert({name: 'A', score: score })
````
???

thearabbit commented:
I tried, It store like this
```js
const score = new Decimal('0.1')
-----------
{
	"_id" : "nCTKu3TZp6mwWzbj2",
	"name" : "A",
	"score" : {
		"s" : 1,
		"e" : -1,
		"d" : [
			1000000
		],
		"name" : "[object Decimal]"
	}
},

zebra-lucky commented:
Yes, all operations are implemented via methods,
String repr can be obtained with https://mikemcl.github.io/decimal.js/#toString

zebra-lucky commented:
You can test how Decimal works at https://mikemcl.github.io/decimal.js/ page.
> The library is incorporated into this page, so it should be available in the console now.

thearabbit commented:
We use this or not???
```js
import { Decimal } from 'meteor/mongo-decimal'
```

thearabbit commented:
I am not about `meteor/mongo-decimal` VS `decimal.js`, which one?

zebra-lucky commented:
Module mongo-decimal stores data in MongoDB as Decimal128.
When data arrive to JS Decimal is same as in decimal.js

zebra-lucky commented:
https://github.com/meteor/meteor/pull/9662/files#diff-b9a0b12de151dca93aa14cd5dc512c3a

thearabbit commented:
It mean that : 
- Install `meteor add mongo-decimal` // for mongodb store
- Install `meteor npm install --save decimal.js` // for js code

### Eg:
- Collection
```js
import { Mongo } from 'meteor/mongo'
import SimpleSchema from 'simpl-schema'

const Students = new Mongo.Collection('demo_students')

Students.schema = new SimpleSchema({
  name: {
    type: String,
  },
  score: {
    type: String,
  },
})

Students.attachSchema(Students.schema)

export default Students
```
- Method
```js
import { Decimal } from 'meteor/mongo-decimal'
....
export const insertStudent = new ValidatedMethod({
  name: 'insertStudent',
  mixins: [CallPromiseMixin],
  validate: null,
  run() {
    if (Meteor.isServer) {
      let data = [
        { name: 'Theara', score: Decimal('0.1') },
        { name: 'Malin', score: Decimal('0.2') },
        { name: 'Maly', score: Decimal('0.3') },
      ]

      data.forEach(it => {
        Students.insert(it)
      })

      return 3
    }
  },
})

export const findStudents = new ValidatedMethod({
  name: 'findStudents',
  mixins: [CallPromiseMixin],
  validate: null,
  run() {
    if (Meteor.isServer) {
      const score = Decimal('0.2')
      return Students.findOne({ score })
    }
  },
})
```

zebra-lucky commented:
> Install meteor add mongo-decimal // for mongodb store
> Install meteor npm install --save decimal.js // for js code
Only ```meteor add mongo-decimal``` is required, npm dependency `decimal.js` is already installed by mongo-decimal. So you only required to ```import { Decimal } from 'meteor/mongo-decimal'``` to use.

Look at package code, it's very simple:
https://github.com/meteor/meteor/tree/devel/packages/non-core/mongo-decimal

I'm not  have time to work with meteor now and not try Decimal with SimpleSchema yet, but think it should work.

There may be a typo:
```
  score: {
    type: String,
  },
```

zebra-lucky commented:
Аnd there is part of code related to store/read in MongoDB Decimal128:

https://github.com/meteor/meteor/pull/9662/files?file-filters%5B%5D=.js#diff-a9d0b23557a439d95f0ceea868fb961b

Thats all.

thearabbit commented:
Now work fine with
```js
score: {
    type: Object,
    blackbox: true
  },
```

thearabbit commented:
Very thanks 👍 💯 .
Could you update `decimal.js` to latest?
(Or I could create it `from this source` on my own project)

zebra-lucky commented:
Sorry, i have no deployed meteor and have limits
on free time to deploy, upgrade, test and make PR.
Maybe someone else can do this.

thearabbit commented:
HI all, now I have problem when fetching data with Aggregate.
1- Meteor collection
```
let data = Collection.find, findOne 
---
doc.decimalField.toNumber() 
// Work fine with Decimal JS Method (Client/Server)
```
2- Use sakulstra:aggregate or Native node driver
```
let data = Collection.aggregate([.......]), Collection.rawCollection().aggregate([.....]) 
---
doc.decimalField.toNumber() 
// Don't work with Decimal JS Method (Client/Server)
```
Please help me…

zebra-lucky commented:
Do you have some traceback?

thearabbit commented:
I don't understand (My English is not good).

zebra-lucky commented:
I replied on https://forums.meteor.com/t/problem-on-sum-of-mongo-aggregate-big-number-01-0-2/48303/15

zebra-lucky commented:
But message awaiting approval

thearabbit commented:
Excuse me, what your name in `Meteor forums`?

zebra-lucky commented:
![Screenshot from 2019-05-23 11-44-08](https://user-images.githubusercontent.com/992125/58238518-1a15c200-7d50-11e9-8577-a02742fa84cd.png)

I'm waiting for approval.
Do you have some errors on meteor console?

thearabbit commented:
OK, thanks this is `PR`!!!!
It don't instance of `Decimal JS` when use` Collection.aggregate([.......])`

zebra-lucky commented:
Can you place more concrete code fragment, please? What is in `[.......]`?

thearabbit commented:
My code
- Collection schema (SimpleSchema)
```js
amount: {
    type: Object, // I tried use `Decimal`, but don't work
    blackbox: true,
  },
```
- Method
```js
import { Decimal } from 'meteor/mongo-decimal'

Meteor.methods({
  insert(){
     return Collection.insert({amount: Decimal(0.123)})
  },
  findOne(){
    return Collection.findOne({})
  },
  aggregate(){
    return Collection.aggregate([{ $limit: 1 }])[0] // sakulstra:aggregate
  }
})
```
- Get data in Client
```js
Meteor.call('findOne', (err, res)=>{
  res.amount.toNumber() // Work fine with Decimal JS method
})
---
Meteor.call('aggregate', (err, res)=>{
  res.amount.toNumber() // don't work with Decimal JS method
})
```

zebra-lucky commented:
I'm try to reproduce problem in a couple of hours.

zebra-lucky commented:
Reproduced problem, returned doc with MongoDB Decimal128:
```I20190523-16:47:50.570(3)? { _id: 'xxNcHJ7LpqLsGtPK4',
I20190523-16:47:50.570(3)?   amount: 
I20190523-16:47:50.571(3)?    Decimal128 {
I20190523-16:47:50.571(3)?      _bsontype: 'Decimal128',
I20190523-16:47:50.571(3)?      bytes: <Buffer 7b 00 00 00 00 00 00 00 00 00 00 00 00 00 3a 30> } }
I20190523-16:47:50.571(3)? undefined```

Looking to code of aggregate.

thearabbit commented:
Yes, the same for me.

zebra-lucky commented:
You can try:

https://atmospherejs.com/zebralucky/aggregate-retyped

https://github.com/zebra-lucky/meteor-aggregate-retyped/

thearabbit commented:
@zebra-lucky, I tried `It work fine` for `Aggregate Retyped`.
How about `Collection.rawCollection()......` of Node Driver???

zebra-lucky commented:
> How about Collection.rawCollection()...... of Node Driver???
MongoDB does not know about Meteor types (lower level abstractions).

zebra-lucky commented:
Maybe additional retyping may be need in other direction, when composing pipeline.

thearabbit commented:
Excuse me @zebra-lucky, I have on Question?
I could use `Decimal` type of of `SimpleSchema & Colleciton2`
```js
import { Mongo } from 'meteor/mongo'
import SimpleSchema from 'simpl-schema'
import { Decimal } from 'meteor/mongo-decimal'

const NumberVal = new Mongo.Collection('demo_numberVal')

NumberVal.schema = new SimpleSchema({
  normalVal: {
    type: Number,
  },
  decimalVal: {
    type: Decimal, // new Decimal()
  },
})

NumberVal.attachSchema(NumberVal.schema)

export default NumberVal
```
But don't work,

zebra-lucky commented:
You previously used blackbox object?

thearabbit commented:
I afraid, It is not the `Decimal` Object?
Ex: Pass argument as Normal Object

zebra-lucky commented:
`simpl-schema` does not know about Decimal also, so use blackbox Object as in previous code fragments:

```js
amount: {
    type: Object, // I tried use `Decimal`, but don't work
    blackbox: true,
  }
```

BTW: Decimal has arbitrary precision, and Decimal128 only 128 bits.
It may hurt when storing overprecised Decimal values in Mongo.
Think I may rewrite code of Decimal to limit precision to 128 bits,
but need more free time for doing it.

thearabbit commented:
HI @zebra-lucky, How to insert `Decimal` with `.rawCollection()`
```js
Collection.rawCollection().insertMany([
  { value: Decimal(0.1) }, // Decimal JS
  { value: NumberDecimal(0.1) }, // Mongo 
  { value: Decimal128(0.1) }, // Node
])
```
But don't work, pl help me...

zebra-lucky commented:
Published version 0.0.2 https://atmospherejs.com/zebralucky/aggregate-retyped

You can import:
```
import {
  replaceTypes, replaceMeteorAtomWithMongo
} from 'meteor/zebralucky:aggregate-retyped';
```

And then look at results:

```
    doc = [
        { value: Decimal(0.1) }
    ];
    console.log(replaceTypes(doc, replaceMeteorAtomWithMongo));
```

Why you need to call rawCollection methods?

thearabbit commented:
Thanks for your reply, sometime I would like to use other Method that don't have in Meteor Collection: `insertMany(), ...`

thearabbit commented:
It is difficult to Store `Mongo Decimal`.
Which one better?, if I store mongo data as `Normal Double`.
And I use `$converter: $toDecimal, $toDouble` in Query Data (.find, findOne, aggregate...) befor calculate it.
```js
Collection.aggregate([
  {
    $project: {
        doublevVal: {$toDecimal: "$doubleVal"}
     }
  },
  {
     $group: {
        _id: null,
        total: { $sum: "$doubleVal"} 
     }
   },
   {
       $project: { total: { $toDouble: "$total"}}
   }
])
```` 
Please advise...

zebra-lucky commented:
> It is difficult to Store Mongo Decimal.
Why not to store Decimal?
I do not understand problem, maybe existing docs contain Number.

thearabbit commented:
Yes, My old project stored `Number Double`. And now I would like to convert to `Decimal`.
But I think that It is not easy by YOUR ADVISE (Insert and Fetching).

thearabbit commented:
Please note, `import {
  replaceTypes, replaceMeteorAtomWithMongo
} from 'meteor/zebralucky:aggregate-retyped';`
in document (Readme.md)

thearabbit commented:
I tried copy this repo to use in Local (Bc can not install `0.0.2`).

Get error for using `replaceTypes, replaceMeteorAtomWithMongo`

```js
{ normalVal: 0.1,
I20190527-10:31:51.257(7)?   decimalVal:
I20190527-10:31:51.257(7)?    Decimal128 {
I20190527-10:31:51.257(7)?      _bsontype: 'Decimal128',
I20190527-10:31:51.258(7)?      bytes: <Buffer 01 00 00 00 00 00 00 00 00 00 00 00 00 00 3e 30> } }
W20190527-10:31:51.333(7)? (STDERR) (node:2893) UnhandledPromiseRejectionWarning: RangeError: Maximum call stack size exceeded
W20190527-10:31:51.334(7)? (STDERR)     at Array.forEach (<anonymous>)
W20190527-10:31:51.334(7)? (STDERR)     at Object.EJSON.clone.v [as clone] (packages/ejson/ejson.js:594:18)
W20190527-10:31:51.334(7)? (STDERR)     at Object.keys.forEach.key (packages/ejson/ejson.js:595:22)
W20190527-10:31:51.334(7)? (STDERR)     at Array.forEach (<anonymous>)
W20190527-10:31:51.335(7)? (STDERR)     at Object.EJSON.clone.v [as clone] (packages/ejson/ejson.js:594:18)
W20190527-10:31:51.335(7)? (STDERR)     at Object.keys.forEach.key (packages/ejson/ejson.js:595:22)
W20190527-10:31:51.335(7)? (STDERR)     at Array.forEach (<anonymous>)
W20190527-10:31:51.335(7)? (STDERR)     at Object.EJSON.clone.v [as clone] (packages/ejson/ejson.js:594:18)
W20190527-10:31:51.335(7)? (STDERR)     at Object.keys.forEach.key (packages/ejson/ejson.js:595:22)
W20190527-10:31:51.336(7)? (STDERR)     at Array.forEach (<anonymous>)
W20190527-10:31:51.336(7)? (STDERR)     at Object.EJSON.clone.v [as clone] (packages/ejson/ejson.js:594:18)
W20190527-10:31:51.336(7)? (STDERR)     at Object.keys.forEach.key (packages/ejson/ejson.js:595:22)
W20190527-10:31:51.336(7)? (STDERR)     at Array.forEach (<anonymous>)
W20190527-10:31:51.336(7)? (STDERR)     at Object.EJSON.clone.v [as clone] (packages/ejson/ejson.js:594:18)
W20190527-10:31:51.337(7)? (STDERR)     at Object.keys.forEach.key (packages/ejson/ejson.js:595:22)
W20190527-10:31:51.337(7)? (STDERR)     at Array.forEach (<anonymous>)
W20190527-10:31:51.337(7)? (STDERR) (node:2893) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 2)
W20190527-10:31:51.337(7)? (STDERR) (node:2893) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
```

thearabbit commented:
Insert code
```js
        let data = {
          normalVal: 0.1,
          // decimalVal: MongoDB.Decimal128.fromString('0.1'),
          decimalVal: Decimal('0.1'),
        } //
        data = replaceTypes(data, replaceMeteorAtomWithMongo)
        console.log(data)

       NumberVal.rawCollection().insertOne(data)
```

thearabbit commented:
Now work fine

jamauro commented:
Thanks for your work on this.

Is it possible to substitute `big.js`  or `bignumber.js` or even `decimal.js-light`? I'd like to use big.js to reduce the bundle size.

I tried to create my own package that wraps `big.js` and creates EJSON type based on the `mongo-decimal` package but when inserting into the mongo db they aren't being stored with `NumberDecimal`.

zebra-lucky commented:
Hi!

It was a long time ago, and the code completely erased from my mind, but let's look at the PR code:
https://github.com/meteor/meteor/pull/9662/files

At the form I've got the message
> Your account associated with this email address has been silenced
Not very evident what this mean

jamauro commented:
Update: I was able to get things working by creating a local `mongo-decimal` package and importing `big.js` in `/mongo-decimal/decimal.js` as `import Decimal from 'big.js'`. I'm not sure why my previous attempt didn't work.

@zebra-lucky do you think it's possible to dynamically config the `mongo-decimal` package to use a decimal library that a user chooses?

zebra-lucky commented:
I'm sorry, I have now limited means to get in the code context and reply.
But a possibility to configure used library seems very logical.
